sql/mysqld.cc

```diff
@@ -6071,38 +6105,198 @@ void handle_connections_sockets()
+#ifdef HAVE_PSI_SOCKET_INTERFACE
+      all_systemd_sockets[systemd_listen_cnt]= (PSI_socket_info) { &all_systemd_key[systemd_listen_cnt], family, PSI_FLAG_GLOBAL };
+#endif
+      setup_fds(systemd_sock[systemd_listen_cnt], socketpeer.ss.ss_family);
+      systemd_fd++;
+      systemd_listen_cnt++;
+    }
+#ifdef HAVE_PSI_SOCKET_INTERFACE
+    all_systemd_sockets[systemd_listen_cnt]= (PSI_socket_info) { &key_socket_client_connection, "client_connection", 0};
+    mysql_socket_register("sql", all_systemd_sockets, systemd_listen_cnt + 1);
+#endif
+    /* print out unused ones */
+    while (systemd_n-- > 0)
+    {
+      socketpeer_len = sizeof(socketpeer);
+      if (getsockname(systemd_fd, (struct sockaddr *) &socketpeer,
+          &socketpeer_len) < 0)
+      {
+        sql_print_error("getsockname for fd %d failed (errno= %d).",
+                        systemd_fd, errno);
+        sd_notifyf(0, "ERRNO=%d", errno);
+        systemd_fd++;
+        continue;
+      }
+      if (socketpeer_len >=sizeof(socketpeer))
+      {
+        sql_print_error("getsockname for fd %d unsufficient space", systemd_fd);
+        sd_notifyf(0, "ERRNO=%d", EINVAL);
+        systemd_fd++;
+        continue;
+      }
+#ifdef HAVE_SYS_UN_H
+      if (socketpeer.ss.ss_family == AF_UNIX)
+      {
+        sql_print_error("Exceeded %d sockets NOT listening on systemd unix socket for %s.",
+                         MAX_LISTEN_SOCKETS, socketpeer.un.sun_path);
+      }
+      else
+#endif
+      if (socketpeer.ss.ss_family == AF_INET ||
+          socketpeer.ss.ss_family == AF_INET6)
+      {
+        sql_print_error("Exceeded %d sockets NOT listening on systemd family %s port %d.",
+              MAX_LISTEN_SOCKETS,
+              socketpeer.ss.ss_family == AF_INET ? "IPv4" : "IPv6",
+              ntohs(socketpeer.ss.ss_family == AF_INET ?
+                       socketpeer.in.sin_port : socketpeer.in6.sin6_port));
+      }
+      systemd_fd++;
+    }
   }
+  else
+  {
+#endif /* HAVE_SYSTEMD */
+    if (mysql_socket_getfd(base_ip_sock) != INVALID_SOCKET)
+    {
+      setup_fds(base_ip_sock, AF_INET);
+      ip_flags = fcntl(mysql_socket_getfd(base_ip_sock), F_GETFL, 0);
+    }
+    if (mysql_socket_getfd(extra_ip_sock) != INVALID_SOCKET)
+    {
+      setup_fds(extra_ip_sock, AF_INET);
+      extra_ip_flags = fcntl(mysql_socket_getfd(extra_ip_sock), F_GETFL, 0);
+    }
 #ifdef HAVE_SYS_UN_H
-  setup_fds(unix_sock);
-  socket_flags=fcntl(mysql_socket_getfd(unix_sock), F_GETFL, 0);
+    setup_fds(unix_sock, AF_UNIX);
+    socket_flags=fcntl(mysql_socket_getfd(unix_sock), F_GETFL, 0);
+#endif
+
+#ifdef HAVE_SYSTEMD
+  }
+
+#ifdef HAVE_POLL
+  if (systemd_listen_cnt == 0 && systemd_n>0)
+  {
+    sd_notify(0, "STATUS=No listening sockets");
+    abort_loop=1;
+  }
+  else
+  {
 #endif
+    sd_notify(0, "READY=1\n"
+               "STATUS=Taking your SQL requests now...");
+#ifdef HAVE_POLL
+  }
+#endif /* HAVE_POLL */
+#endif /* HAVE_SYSTEMD */
```

Comment:
> Socket activation is used be able to start services in parallel while automatically introducing dependencies between them.

Yes, and it's also used to isolate privilege dropping, pass listening sockets into containers, and other cases that have nothing to do with start-on-demand.

> So, say, mysqld and mythtv are started in parallel on boot. And then mythtv is up and tries to access the database — it can as well time out before the database is ready.

So make MythTV be `Requires=mysqld.service` and be `After=mysqld.service`. Assuming `mysqld.service` specifies that it depends on its socket unit (which is necessary if bringing it up without its socket being passed in would break things), everything will work fine no matter how long the database takes to come online. systemd will use the "notify" data to determine when `mysqld.service` is ready and will only bring MythTV online at that point.